scripts/safe_template_checker.py

```diff
@@ -0,0 +1,498 @@
+
+
+def _get_line_breaks(self, file_string):
+    line_breaks = [0]
+    index = 0
+    while True:
+        index = file_string.find('\n', index)
+        if index < 0:
+            break
+        index += 1
+        line_breaks.append(index)
+    return line_breaks
+
+
+def _get_line_number(self, line_breaks, index):
+    current_line_number = 0
+    for line_break_index in line_breaks:
+        if line_break_index <= index:
+            current_line_number += 1
+        else:
+            break
+    return current_line_number
+
+
+def _get_line(self, file_string, line_breaks, line_number):
+    start_index = line_breaks[line_number - 1]
+    if len(line_breaks) == line_number:
+        line = file_string[start_index:]
+    else:
+        end_index = line_breaks[line_number]
+        line = file_string[start_index:end_index - 1]
+    return line.encode(encoding='utf-8')
+
+
+def _get_column_number(self, line_breaks, line_number, index):
+    start_index = line_breaks[line_number - 1]
+    column = index - start_index + 1
+    return column
+
+
+class Rules(Enum):
+    mako_missing_default = ('mako-missing-default', 'The default page directive with h filter is missing.')
+    mako_unparsable_expression = ('mako-unparsable-expression', 'The expression could not be properly parsed.')
+    mako_unwanted_html_filter = ('mako-unwanted-html-filter', 'Remove explicit h filters when it is provided by the page directive.')
+    mako_invalid_html_filter = ('mako-invalid-html-filter', 'The expression is using an invalid filter in an HTML context.')
+    mako_invalid_js_filter = ('mako-invalid-js-filter', 'The expression is using an invalid filter in a JavaScript context.')
+    mako_js_string_missing_quotes = ('mako-js-string-missing-quotes', 'An expression using the js_escape_string filter must have surrounding quotes.')
+
+    underscore_not_escaped = ('underscore-not-escaped', 'Expressions should be escaped using <%- expression %>.')
+
+    def __init__(self, rule_id, rule_summary):
+        self.rule_id = rule_id
+        self.rule_summary = rule_summary
+
+
+class BrokenRule(object):
+
+    def __init__(self, rule):
+        self.rule = rule
+        self.full_path = ''
+
+    def prepare_results(self, full_path, file_string, line_breaks):
+        self.full_path = full_path
+
+    def print_results(self, options):
+        print "{}: {}".format(self.full_path, self.rule.rule_id)
+
+
+class BrokenExpressionRule(BrokenRule):
+
+    def __init__(self, rule, expression):
+        super(BrokenExpressionRule, self).__init__(rule)
+        self.expression = expression
+        self.start_line = 0
+        self.start_column = 0
+        self.end_line = 0
+        self.end_column = 0
+        self.lines = []
+
+    def prepare_results(self, full_path, file_string, line_breaks):
+        self.full_path = full_path
+        start_index = self.expression['start_index']
+        self.start_line = _get_line_number(self, line_breaks, start_index)
```

Comment:
All of this logic makes me think that lines 42-71 should be a class called IndexedSourceFile or something.